Bottom: e84b4fd3acbb89d94bc378e37c404462fa23cd00
Top:    0a7024b7ae9a9367a5076b9de237cd5b3626804c
Author: Jason Wang <jasowang@redhat.com>
Date:   2017-08-17 16:59:20 +0800

tun: reserve extra headroom only when XDP is set

Signed-off-by: Jason Wang <jasowang@redhat.com>


---

diff --git a/drivers/net/tun.c b/drivers/net/tun.c
index 19cbbbb..674a8e2 100644
--- a/drivers/net/tun.c
+++ b/drivers/net/tun.c
@@ -1272,14 +1272,21 @@ static struct sk_buff *tun_build_skb(struct tun_struct *tun,
 	struct page_frag *alloc_frag = &current->task_frag;
 	struct sk_buff *skb;
 	struct bpf_prog *xdp_prog;
-	int buflen = SKB_DATA_ALIGN(len + TUN_RX_PAD) +
-		     SKB_DATA_ALIGN(sizeof(struct skb_shared_info));
+	int buflen = SKB_DATA_ALIGN(sizeof(struct skb_shared_info));
 	unsigned int delta = 0;
 	char *buf;
 	size_t copied;
 	bool xdp_xmit = false;
 	int err;
 
+	rcu_read_lock();
+	xdp_prog = rcu_dereference(tun->xdp_prog);
+	if (xdp_prog)
+		buflen += SKB_DATA_ALIGN(len + TUN_RX_PAD);
+	else
+		buflen += SKB_DATA_ALIGN(len);
+	rcu_read_unlock();
+
 	if (unlikely(!skb_page_frag_refill(buflen, alloc_frag, GFP_KERNEL)))
 		return ERR_PTR(-ENOMEM);
 
@@ -1290,7 +1297,11 @@ static struct sk_buff *tun_build_skb(struct tun_struct *tun,
 	if (copied != len)
 		return ERR_PTR(-EFAULT);
 
-	if (hdr->gso_type)
+	/* There's a small window that XDP may be set after the check
+	 * of xdp_prog above, this should be rare and for simplicity
+	 * we do XDP on skb in case the headroom is not enough.
+	 */
+	if (hdr->gso_type || !xdp_prog)
 		*generic_xdp = 1;
 	else
 		*generic_xdp = 0;
